odl/discr/discr_ops.py

```diff
@@ -130,110 +144,212 @@ def finite_diff(f, out=None, axis=0, dx=1.0, edge_order=2,
+
+    if edge_order is None:
+        if method == 'central':
+            edge_order = 2
+        else:
+            edge_order = 1
+    else:
+        if edge_order not in [1, 2]:
+            raise ValueError('edge order ({0}) not valid'.format(edge_order))
+
+    if f_arr.shape[axis] == 2 and edge_order == 2:
+        raise ValueError('shape ({0}) of array to small to use one-sided '
+                         'second-order accurate edges'.format(f_arr.shape))
+
+    if padding is not None:
+        if method == 'central' and edge_order == 1:
+            raise ValueError(
+                'central differences with padding cannot be used with '
+                'first-order accurate edges')
+        if method in ('forward', 'backward') and edge_order == 2:
+            raise ValueError(
+                '{} differences with padding use the same '
+                'accuracy on edges as in the interior. Edge-order accuracy '
+                'can only be triggered without padding for one-sided '
+                'differences at edges.'.format(method))
+
     # create slice objects: initially all are [:, :, ..., :]
+
     # current slice
     slice_out = [slice(None)] * ndim
+
     # slices used to calculate finite differences
     slice_node1 = [slice(None)] * ndim
     slice_node2 = [slice(None)] * ndim
     slice_node3 = [slice(None)] * ndim
 
+    # Interior
+
     if method == 'central':
-        # Numerical differentiation: 2nd order interior
+        # 2nd order interior
         slice_out[axis] = slice(1, -1)
         slice_node1[axis] = slice(2, None)
         slice_node2[axis] = slice(None, -2)
         # 1D equivalent: out[1:-1] = (f[2:] - f[:-2])/2.0
-        np.subtract(f_data[slice_node1], f_data[slice_node2], out[slice_out])
+        np.subtract(f_arr[slice_node1], f_arr[slice_node2], out[slice_out])
         out[slice_out] /= 2.0
+
     elif method == 'forward':
-        # Numerical differentiation: 1nd order interior
+        # 1st order interior
         slice_out[axis] = slice(1, -1)
         slice_node1[axis] = slice(2, None)
         slice_node2[axis] = slice(1, -1)
         # 1D equivalent: out[1:-1] = (f[2:] - f[1:-1])
-        np.subtract(f_data[slice_node1], f_data[slice_node2], out[slice_out])
+        np.subtract(f_arr[slice_node1], f_arr[slice_node2], out[slice_out])
+
     elif method == 'backward':
-        # Numerical differentiation: 1nd order interior
+        # 1st order interior
         slice_out[axis] = slice(1, -1)
         slice_node1[axis] = slice(1, -1)
         slice_node2[axis] = slice(None, -2)
         # 1D equivalent: out[1:-1] = (f[1:-1] - f[:-2])
-        np.subtract(f_data[slice_node1], f_data[slice_node2], out[slice_out])
+        np.subtract(f_arr[slice_node1], f_arr[slice_node2], out[slice_out])
 
-    if zero_padding:
-        # Assume zeros for indices outside the domain of `f`
+    # Boundaries
+
+    if isinstance(padding, float):
```

Comment:
Also not so great to use this internally. Better save the constant in a new variable `padding_const` or so and set `padding = 'constant'` above. Then this becomes just another case.